src/state_machine.zig

```diff
@@ -552,19 +537,18 @@ pub fn StateMachineType(
                 },
                 inline else => |comptime_operation| {
                     const event_size = @sizeOf(Event(comptime_operation));
-                    const batch_max = comptime constants.batch_max.operation_batch_max(
-                        comptime_operation,
-                    );
-
                     comptime assert(event_size > 0);
-                    comptime assert(batch_max > 0);
+
+                    const batch_limit: u32 =
+                        operation_batch_max(comptime_operation, self.batch_size_limit);
+                    assert(batch_limit > 0);
 
                     // Clients do not validate batch size == 0,
                     // and even the simulator can generate requests with no events.
                     maybe(input.len == 0);
 
                     if (input.len % event_size != 0) return false;
-                    if (input.len > batch_max * event_size) return false;
+                    if (input.len > batch_limit * event_size) return false;
                 },
             }
 
```

Comment:
assert `input.len < self.batch_size_limit`, as different switch branches use diferent logic to enforce that invariant.

Reply:
Ah, I see that we do that at the call-site, but it still feels worthwhile to do this inside, so that one can AND reasoning from each branch.